Add conditions evaluation operation.

pronebird opened the pull request:
Previously condition evaluation would happen on Operation level.
However due to bug discovered in NSOperation we cannot safely manipulate NSOperation.isReady.

Therefore we have to wrap condition evalutor into operation and make it dependency for main operation.
Evaluation errors are then passed back to main operation before its execution.

Relevant discussions:
1. danthorpe/Operations#175
2. Kabal#51

Commit:
Add conditions evaluation operation.
Previously condition evaluation would happen on Operation level.
However due to bug discovered in NSOperation we cannot safely manipulate NSOperation.isReady.

Therefore we have to wrap condition evalutor into operation and make it dependency for main operation.
Evaluation errors are then passed back to main operation before its execution.

Relevant discussions:

1. danthorpe/Operations#175
2. Kabal#51

pronebird commented:
So I found that the problem was in `NSOperation.isReady`. My guess is `NSOperation` has some inconsistency in its logic and it does not always consult subclass regarding operation readiness. Operation does not start after shifting into Ready state as a result. 

This PR only manipulates `isExecuting` and `isFinished` states. `isReady` is entirely managed by `NSOperation` via dependencies mechanism which works fine.

 That's why conditions evaluator is wrapped in operation. Conditions Evaluation operation now waits for all dependencies from conditions to finish before running evaluator. Operation has dependency set for evaluator's operation. 

operation -> evaluator -> dependencies from conditions

`OPOperation.internalErrors` is filled up with errors just before evaluation operation finishes execution.

This allows `OPOperation` to start naturally like before and gracefully handle errors coming from conditions. `willEnqueue` is not used anymore and is a no-op.

This change should have no impact on apps using Operative. However, it's BIG. I wrote two stress tests to make sure that there is no crash or stale operation. I manually checked if conditions work properly, location condition from Example app works just fine, evaluation happens after permission alert, just like before.

I would appreciate a lot of testing before we merge.

pronebird commented on Example/Tests/StressTests.m:
```diff
@@ -0,0 +1,203 @@
+    
+    [self cycleOperationsWithHollowCondition:1 operationQueue:queue expectation:expectation];
+    
+    // time for cofee: give it 10 minutes
```
oh jesus, typo.

Commit:
Fix typo

pronebird commented:
Don't rush to merge, I am still adding tests and checking mutual exclusivity as it seems that conditions have to be mutually exclusive too in order to run in sequence for mutually exclusive operations.

colinmcardell commented:
@pronebird Wow. Lots of changes. I'm concerned about the idea of completely deviating from the state logic of Advanced NSOperations, although it has been mentioned in discussions elsewhere that Operative is not intended to necessarily directly reflect Apples implementation. @Kabal thoughts?

FWIW, I pulled https://github.com/inspace-io/INSOperationsKit and built up a test case identical to what you are testing (dependency chain + mutual exclusivity). I can fire off and complete a test with 500 - 5000 operations (50,000 took forever to getting the ball rolling so I stopped it, seemed excessive although there was nothing leading me to believe that it wouldn't successfully complete). The logic around state and readiness within INSOperationsKit seems to be _pretty much_ identical to Advanced NSOperations...

I do think it is clear as you stated here https://github.com/Kabal/Operative/pull/43#issuecomment-209141873 that we should be pulling the KVO `-will...` `-did...` calls out of the sync lock.

pronebird commented:
@colinmcardell 

I've been trying to fix that for entire week to make sure I commit as less changes as possible without any success. This is the last resort for me.

I made a stripped down version of OPOperation with test case that reproduces the same problem we have in Operative and Operations: 

https://github.com/pronebird/NSOperationReproducingRaceCondition. 

Also, I was able to reproduce the same problem in Earthquakes app:

https://gist.github.com/pronebird/6ccf4e67030f5fdffe65c941c736181c

I don't think the problem is on our side, I think something is fundamentally broken inside of NSOperation that prevents us from overriding isReady.

Please do not merge this PR yet, I have to solve mutual exclusivity issue. Apparently evaluator operation should be mutually exclusive too when using mutual exclusivity on operations... WIP.

But good to speak about this change sooner than later.

Commit:
- Condition evaluator nor its dependencies should not run before previous mutually exclusive operation finished
- Add tests for operation evaluation

pronebird commented:
It's not elegant, but I need it working. I hope that you two can help to polish it, rethink or merge. Last part was to make sure that condition evaluation operation does not run before previous mutually exclusive op. I've implemented this by adding dependency between previous op and evaluator. This is a small bit of change in ExclusivityController.

Commit:
Prevent use of group operations which can cause deadlock when using produceOperation from within condition dependency:

- Remove conditionEvaluationOperation from OPOperation
- OPExclusivityController.addOperation returns a set of previous mutually exclusive ops
- Wire up evaluator operation and all dependencies in OPOperationQueue

Commit:
Add tests to verify that queue does not deadlock when condition dependencies produce mutually exclusive operations with the same categories as relevant operations

Commit:
Pass evaluation errors back to operation

Commit:
Previously we would finish operation if it was executing during cancellation.
